src/glue_launcher_lambda/glue_launcher.py

```diff
@@ -0,0 +1,517 @@
+        ]
+
+    if "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISSING_IMPORT" in os.environ:
+        _args.manifest_s3_input_parquet_location_missing_import = os.environ[
+            "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISSING_IMPORT"
+        ]
+
+    if "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISSING_EXPORT" in os.environ:
+        _args.manifest_s3_input_parquet_location_missing_export = os.environ[
+            "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISSING_EXPORT"
+        ]
+
+    if "MANIFEST_S3_INPUT_PARQUET_LOCATION_COUNTS" in os.environ:
+        _args.manifest_s3_input_parquet_location_counts = os.environ[
+            "MANIFEST_S3_INPUT_PARQUET_LOCATION_COUNTS"
+        ]
+
+    if "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISMATCHED_TIMESTAMPS" in os.environ:
+        _args.manifest_s3_input_parquet_location_mismatched_timestamps = os.environ[
+            "MANIFEST_S3_INPUT_PARQUET_LOCATION_MISMATCHED_TIMESTAMPS"
+        ]
+
+    if "MANIFEST_S3_OUTPUT_LOCATION" in os.environ:
+        _args.manifest_s3_output_location = os.environ["MANIFEST_S3_OUTPUT_LOCATION"]
+
+    return _args
+
+
+def get_escaped_json_string(json_string):
+    try:
+        escaped_string = json.dumps(json.dumps(json_string))
+    except:
+        escaped_string = json.dumps(json_string)
+
+    return escaped_string
+
+
+def get_and_validate_job_details(message):
+    """Get the job name from the event.
+    Arguments:
+        event (dict): The event
+    """
+
+    dumped_message = get_escaped_json_string(message)
+    logger.info(f'Validating message", "message_details": {dumped_message}')
+
+    if "detail" not in message:
+        raise KeyError("Message contains no 'detail' key")
+
+    detail_dict = message["detail"]
+    required_keys = [JOB_NAME_KEY, JOB_STATUS_KEY, JOB_QUEUE_KEY]
+
+    for required_key in required_keys:
+        if required_key not in detail_dict:
+            error_string = f"Details dict contains no '{required_key}' key"
+            raise KeyError(error_string)
+
+    logger.info(
+        f'Message has been validated", "message_details": {dumped_message}, "job_queue": "{detail_dict[JOB_QUEUE_KEY]}", '
+        + f'"job_name": "{detail_dict[JOB_NAME_KEY]}", "job_status": "{detail_dict[JOB_STATUS_KEY]}'
+    )
+
+    return detail_dict
+
+
+def generate_ms_epoch_from_timestamp(formatted_timestamp_string, minutes_to_add=0):
+    """Returns the 1970 epoch as a number from the given timestamp.
+
+    Keyword arguments:
+    timestamp_string -- the timestamp as a string formatted to %Y-%m-%dT%H:%M:%S.%f%z
+    minutes_to_add -- if any minutes are to be added to the time, set to greater than 0
+    """
+    timestamp = datetime.strptime(formatted_timestamp_string, "%Y-%m-%dT%H:%M:%S.%f")
+    if minutes_to_add > 0:
+        timestamp = timestamp + timedelta(minutes=minutes_to_add)
+
+    return int((timestamp - datetime(1970, 1, 1)).total_seconds() * 1000.0)
+
+
+def get_batch_client():
+    return boto3.client("batch")
```

Comment:
Add retries to all these clients. See the ADG python code for the generate_dataset step. I added retries in there, you can copy that.

Reply:
Example -> https://github.com/dwp/aws-analytical-dataset-generation/blob/master/steps/generate_dataset_from_htme.py#L625